nixos/tests/nvidia-container-toolkit.nix

```diff
@@ -0,0 +1,158 @@
+        config = {
+          Cmd = [ "${testCDIScript}/bin/test-cdi" ];
+          Env = [
+            "LD_LIBRARY_PATH=${unfreeAndInsecurePkgs.linuxPackages.nvidia_x11}/lib:${lib.getLib unfreeAndInsecurePkgs.openssl_1_1}/lib"
+          ];
+        };
+        copyToRoot = (
+          with pkgs.dockerTools;
+          [
+            usrBinEnv
+            binSh
+          ]
+        );
+      };
+  in
+  {
+    name = "nvidia-container-toolkit";
+    meta = with lib.maintainers; {
+      maintainers = [ ereslibre ];
+    };
+    nodes = {
+      no-nvidia-gpus = {
+        environment.systemPackages = with pkgs; [ jq ];
+        hardware.nvidia-container-toolkit.enable = true;
+        nixpkgs.config.allowUnfree = true;
+      };
+      nvidia-one-gpu =
+        { pkgs, ... }:
+        let
+          emptyCDISpec = ''
+            #! ${pkgs.runtimeShell}
+            cat <<CDI_DOCUMENT
+              {
+                "cdiVersion": "0.5.0",
+                "kind": "nvidia.com/gpu",
+                "devices": [
+                  {
+                    "name": "all",
+                    "containerEdits": {
+                      "deviceNodes": [
+                        {
+                          "path": "/dev/urandom"
+                        }
+                      ],
+                      "hooks": [],
+                      "mounts": []
+                    }
+                  }
+                ],
+                "containerEdits": {
+                  "deviceNodes": [],
+                  "hooks": [],
+                  "mounts": []
+                }
+              }
+            CDI_DOCUMENT
+          '';
+        in
+        {
+          virtualisation.diskSize = 10240;
+          environment.systemPackages = with pkgs; [
+            jq
+            podman
+          ];
+          hardware = {
+            nvidia-container-toolkit = {
+              enable = true;
+              package = pkgs.stdenv.mkDerivation {
+                name = "nvidia-ctk-dummy";
+                version = "1.0.0";
+                phases = "installPhase";
+                installPhase = ''
+                  mkdir -p $out/bin
+                  cat <<EOF > $out/bin/nvidia-ctk
+                    ${emptyCDISpec}
+                  EOF
+                  chmod +x $out/bin/nvidia-ctk
+                '';
+              };
+            };
+            opengl.enable = true;
+          };
+          nixpkgs.config.allowUnfree = true;
+          services.xserver.videoDrivers = [ "nvidia" ];
+          virtualisation.containers.enable = true;
+        };
+    };
+    testScript = ''
+      start_all()
+
+      with subtest("Generate the CDI spec (empty) for a machine with no Nvidia GPU"):
+        no_nvidia_gpus.wait_for_unit("nvidia-container-toolkit-cdi-generator.service")
+        no_nvidia_gpus.succeed("cat /var/run/cdi/nvidia-container-toolkit.json | jq")
+
+      with subtest("Generate the CDI spec for a machine with an Nvidia GPU"):
```

Comment:
Looking at the test, I suppose what it ensures is that `podman` as ran with `--device=...` does read the spec from the expected location... I didn't quite get that idea from the test name

Reply:
Yes exactly, and that the JSON is valid etc. I'll name the test in a more descriptive way.